group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -874,4 +1072,1338 @@ public void replay(
+        GenericGroupMember member = new GenericGroupMember(
+            memberId,
+            groupInstanceId,
+            context.clientId(),
+            context.clientAddress().toString(),
+            request.rebalanceTimeoutMs(),
+            request.sessionTimeoutMs(),
+            request.protocolType(),
+            request.protocols()
+        );
+
+        member.setIsNew(true);
+
+        // Update the newMemberAdded flag to indicate that the initial rebalance can be further delayed
+        if (group.isInState(PREPARING_REBALANCE) && group.generationId() == 0) {
+            group.setNewMemberAdded(true);
+        }
+        
+        group.add(member, responseFuture);
+
+        // The session timeout does not affect new members since they do not have their memberId and
+        // cannot send heartbeats. Furthermore, we cannot detect disconnects because sockets are muted
+        // while the JoinGroup request is parked. If the client does disconnect (e.g. because of a request
+        // timeout during a long rebalance), they may simply retry which will lead to a lot of defunct
+        // members in the rebalance. To prevent this going on indefinitely, we time out JoinGroup requests
+        // for new members. If the new member is still there, we expect it to retry.
+        rescheduleGenericGroupMemberHeartbeat(group, member, genericGroupNewMemberJoinTimeoutMs);
+
+        return maybePrepareRebalance(group, "Adding new member " + memberId + " with group instance id " +
+            request.groupInstanceId() + "; client reason: " + joinReason);
+    }
+
+    /**
+     * Prepare a rebalance if the group is in a valid state.
+     *
+     * @param group           The group to rebalance.
+     * @param reason          The reason for the rebalance.
+     *
+     * @return The coordinator result that will be appended to the log.
+     */
+    private CoordinatorResult<CompletableFuture<Void>, Record> maybePrepareRebalance(
+        GenericGroup group,
+        String reason
+    ) {
+        if (group.canRebalance()) {
+            return prepareRebalance(group, reason);
+        }
+
+        return EMPTY_RESULT;
+    }
+
+    /**
+     * Prepare a rebalance.
+     *
+     * @param group           The group to rebalance.
+     * @param reason          The reason for the rebalance.
+     *
+     * @return The coordinator result that will be appended to the log.
+     */
+    CoordinatorResult<CompletableFuture<Void>, Record> prepareRebalance(
+        GenericGroup group,
+        String reason
+    ) {
+        // If any members are awaiting sync, cancel their request and have them rejoin.
+        if (group.isInState(COMPLETING_REBALANCE)) {
+            resetAndPropagateAssignmentWithError(group, Errors.REBALANCE_IN_PROGRESS);
+        }
+
+        // If a sync expiration is pending, cancel it.
+        removeSyncExpiration(group);
+
+        boolean isInitialRebalance = group.generationId() == 0;
```

Comment:
I thought the "initial" rebalance only applied to when the group is first created from https://github.com/apache/kafka/pull/2758 but according to https://cwiki.apache.org/confluence/display/KAFKA/KIP-134%3A+Delay+initial+consumer+group+rebalance it looks like we want this for an empty group as well

> Adding a configurable delay to the initial rebalance of a new, or empty, consumer group [...]

So, i will revert this change.

However, it's awkward because here we consider "initial rebalance" to be an empty group. But a different part of the code checks generation id == 0. maybe it's because we don't know the previous state

from GroupCoordinator#addMemberAndRebalance:
```
    // update the newMemberAdded flag to indicate that the join group can be further delayed
    if (group.is(PreparingRebalance) && group.generationId == 0)
      group.newMemberAdded = true
```

I think we need to revert the change and include `initialRebalanceDelayMs` to the GenericGroup object and rely on that to check whether the group is undergoing an initial rebalance.

Reply:
added a previousState to GenericGroup. we will rely on this instead to confirm that a group is undergoing an initial rebalance (previous state == empty)